# Copyright (c) 2013 Per Unneberg
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.
"""
The haloplex pipeline implements a workflow recommended by Agilent technologies. 

It adds the following custom tasks:

1. :class:`.RawUnifiedGenotyper`
2. :class:`.VariantHaloFiltration`
3. :class:`.HaloPlexUnifiedGenotyperAlleles`
4. :class:`.CombineAllVariants`

The main pipeline tasks are 

1. :class:`.HaloPlex`
2. :class:`.HaloPlexSummary`
3. :class:`.HaloPlexCombine`

Running only HaloPlexSummary or HaloPlexCombine should invoke all dependencies in HaloPlex.
However, for many samples (>20), it is advisable to first run HaloPlex
with smaller number of samples in batches.


Calling via ratatosk_run.py
----------------------------

.. code-block:: text

   ratatosk_run.py HaloPlex --indir inputdir --custom-config custom_config_file.yaml
   ratatosk_run.py HaloPlexSummary --indir inputdir --custom-config custom_config_file.yaml
   ratatosk_run.py HaloPlexCombine --indir inputdir --custom-config custom_config_file.yaml


Classes
-------
"""

import luigi
import os
from ratatosk import backend
from ratatosk.job import PipelineTask, JobTask, JobWrapperTask, PrintConfig
from ratatosk.utils import make_fastq_links, rreplace, fullclassname
from ratatosk.lib.tools.gatk import VariantEval, UnifiedGenotyper, UnifiedGenotyperAlleles, VariantFiltration, CombineVariants
from ratatosk.lib.variation.tabix import Bgzip
from ratatosk.log import get_logger
import ratatosk.lib.tools.samtools

logger = get_logger()

class CombineAllVariants(CombineVariants):
    """Combine all variants generated by :class:`.HaloPlexUnifiedGenotyperAlleles`
    """
    _config_section = "ratatosk.lib.tools.gatk"

    def requires(self):
        targets = backend.__global_vars__["targets"]
        out_targets = list(set(["{}.{}".format(x.prefix("sample"), "trimmed.sync.sort.merge.realign.recal.clip.filtered-genotype.vcf") for x in targets]))
        return [HaloPlexUnifiedGenotyperAlleles(target=tgt, outdir=os.path.dirname(self.target)) for tgt in out_targets]

class HaloPlexUnifiedGenotyperAlleles(UnifiedGenotyperAlleles):
    """ Temporary class that resolves the issue of calling
    :class:`CombineVariants <ratatosk.lib.tools.gatk.CombineVariants>`
    with a predefined target name.
    """
    parent_task = luigi.Parameter(default=("ratatosk.lib.tools.gatk.ClipReads",
                                           "ratatosk.lib.tools.gatk.CombineVariants"), is_list=True)
    _config_section = "ratatosk.lib.tools.gatk"
    outdir = luigi.Parameter(description="Where analysis takes place", default=None)
    def requires(self):
        """Task requirements. In many cases this is a single source
        whose name can be generated following the code below, and
        therefore doesn't need reimplementation in the subclasses."""
        bamcls = self.parent()[0]
        indexcls = ratatosk.lib.tools.samtools.Index
        return [bamcls(target=self.source()[0])]  + [CombineVariants(target=os.path.join(self.outdir, "CombinedVariants.vcf"))] + [indexcls(target=rreplace(self.source()[0], bamcls().sfx(), indexcls().sfx(), 1), parent_task=fullclassname(bamcls))]


class RawUnifiedGenotyper(UnifiedGenotyper):
    """
    RawUnifiedGenotyper is a variant calling class done on merged data
    to generate a list of raw candidates around which realignment is
    done.
    """
    _config_section = "ratatosk.lib.tools.gatk"
    parent_task = luigi.Parameter(default=("ratatosk.lib.tools.picard.MergeSamFiles", "ratatosk.lib.tools.picard.PicardMetrics"), is_list=True)
    options = luigi.Parameter(default=("-stand_call_conf 30.0 -stand_emit_conf 10.0  --downsample_to_coverage 30 --output_mode EMIT_VARIANTS_ONLY -glm BOTH",), is_list=True)
    label = luigi.Parameter(default=".BOTH.raw")

class VariantHaloFiltration(VariantFiltration):
    """VariantHaloFiltration implements settings tailored for
    filtering haloplex variant calls

    """
    _config_section = "ratatosk.lib.tools.gatk"
    # Options from Halo
    options = luigi.Parameter(default=('--clusterWindowSize 10 --clusterSize 3 --filterExpression "MQ0 >= 4 && ((MQ0 / (1.0 * DP)) > 0.1)" --filterName "HARD_TO_VALIDATE" --filterExpression "DP < 10" --filterName "LowCoverage" --filterExpression "QUAL < 30.0" --filterName "VeryLowQual" --filterExpression "QUAL > 30.0 && QUAL < 50.0" --filterName "LowQual" --filterExpression "QD < 1.5" --filterName "LowQD"',), is_list=True)


class HaloPipeline(PipelineTask):
    """Main pipeline task for HaloPlex. All HaloPlex pipeline tasks
    subclass this task."""
    # Weird: after subclassing job.PipelineTask, not having a default
    # here throws an incomprehensible error
    indir = luigi.Parameter(description="Where raw data lives", default=None)
    outdir = luigi.Parameter(description="Where analysis takes place", default=None)
    target_generator_file = luigi.Parameter(description="Target generator file name", default=None, is_global=True)
    sample = luigi.Parameter(default=[], description="Samples to process.", is_list=True)
    flowcell = luigi.Parameter(default=[], description="Flowcells to process.", is_list=True)
    lane = luigi.Parameter(default=[], description="Lanes to process.", is_list=True)
    # Hard-code this for now - would like to calculate on the fly so that
    # implicit naming is unnecessary
    final_target_suffix = "trimmed.sync.sort.merge.realign.recal.clip.filtered.eval_metrics"    
    targets = []

    def _setup(self):
        # List requirements for completion, consisting of classes above
        if self.indir is None:
            logger.error("Need input directory to run")
            self.targets = []
        if self.outdir is None:
            self.outdir = self.indir
        self.targets = [tgt for tgt in self.target_iterator()]
        if self.outdir != self.indir and self.targets:
            self.targets = make_fastq_links(self.targets, self.indir, self.outdir)
        # Finally register targets in backend
        backend.__global_vars__["targets"] = self.targets

class HaloPlex(HaloPipeline):
    """Main HaloPlex pipeline. Performs the following steps:

    1. Adapter trimming
    2. Alignment
    3. Raw genotyping, followed by indel realignment and base recalibration
    4. Genotyping and filtering

    """
    def requires(self):
        self._setup()
        if not self.targets:
            return []
        variant_targets = ["{}.{}".format(x.prefix("sample"), self.final_target_suffix) for x in self.targets]
        return [VariantEval(target=tgt) for tgt in variant_targets]

class HaloBgzip(Bgzip):
    _config_section = "ratatosk.lib.variation.tabix"
    parent_task = luigi.Parameter(default=("ratatosk.lib.variation.htslib.VcfMerge", ), is_list=True)

class HaloPlexSummary(HaloPipeline):
    """Merges and indexes vcf files with :program:`tabix` tools. NB:
    samples are *not* genotyped at the 'union' of all candidate
    positions. See :class:`.HaloPlexCombine` for this functionality.
    """

    def requires(self):
        self._setup()
        return [HaloBgzip(target=os.path.join(self.outdir, "all.vcfmerge.vcf.gz"))]

class HaloPlexCombine(HaloPipeline):
    """Combine variants from samples by:

    1. generating a master vcf with :class:`CombineVariants <ratatosk.lib.tools.gatk.CombineVariants>`
    2. genotyping samples at positions given by master vcf
    3. combining genotyped samples with :class:`.CombineAllVariants`
    """
    def requires(self):
        self._setup()
        return [CombineAllVariants(target=os.path.join(self.outdir, "CombinedVariantsAll.vcf"))]
